### PR TITLE
Max retries support for existing restart strategies

### PR DESCRIPTION
**Pull request by Relrin**
The following pull request contains the changes for supporting max retries for the failed actors. It covers the #105 and the #146 issues for the bastion project.

By default the restart strategy doesn't have any limits in restating the failed actor, but can be overridden via the special call (which is the `.with_max_retries`).

##### Checklist

- [x] tests are passing with `cargo test`. 
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message is clear
- [x] added an example of the usage

### CODE REVIEW

**Comment by o0Ignition0o** [comment]
Thanks for your pull request!

It looks great overall, And I'm really pleased to see the new example and the docs update.
I won't be able to look at it before tomorrow, but maybe @vertexclique will :)

**Comment by Relrin** [comment]
This PR is still in progress and restricted by the following issue:
The following [piece of code](https://github.com/bastion-rs/bastion/blob/f3d3ed0c2b004b9395ab782e103137e0791e7cfc/bastion/src/supervisor.rs#L944-L1007) contains the separate logic for restoring the single actor and the range. I'd like to use the single call for restoring a failed actor. Any ideas how to reorganize signature of the `async fn restart(&mut self, range: RangeFrom<usize>)` method, so that I could do write as: 
```rust
match self.strategy {
            SupervisionStrategy::OneForOne => {
                // expected call 
                self.restart(id..id+1).await; 
            }
            SupervisionStrategy::OneForAll => {
                self.restart(0..).await;


                self.stopped.shrink_to_fit();
                self.killed.shrink_to_fit();
            }
            SupervisionStrategy::RestForOne => {
                let (start, _, _) = self.launched.get(&id).ok_or(())?;
                let start = *start;

                self.restart(start..).await;
            }
        }
```

**Comment by o0Ignition0o** [comment]
That’s a pretty tricky question! Off the top of my head the function could take an into iterator over bastionId so we could invoke it with std::iter::once(id) or with a range, but I’m not quite sure it’s worth the hassle.

(Sorry I’m not at the computer right now so I can’t develop much more on that -pun intended-)

**Comment by vertexclique** [comment]
One more idea would be:
* Seek in the range
* Use `splice` to extract the chunk that contains our one and the only.
* Yield to your method with dereffed chunk and use 0 on that (or directly use `first`).

Will take a look @ home again.

**Comment by Relrin** [comment]
> One more idea would be:
> 
> * Seek in the range
> * Use `splice` to extract the chunk that contains our one and the only.
> * Yield to your method with dereffed chunk and use 0 on that (or directly use `first`).

I also thinking about the following solution:
```rust
async fn recover(&mut self, id: BastionId) -> Result<(), ()> {
        debug!(
            "Supervisor({}): Recovering using strategy: {:?}",
            self.id(),
            self.strategy
        );
        match self.strategy {
            SupervisionStrategy::OneForOne => {
                let (index, _, _) = self.launched.remove(&id).ok_or(())?;
                let value = self.order.remove(index);
                self.order.push(value);
                let start = self.order.len();
                self.restart(start..).await;
            }
            SupervisionStrategy::OneForAll => {
                self.restart(0..).await;

                // TODO: should be empty
                self.stopped.shrink_to_fit();
                self.killed.shrink_to_fit();
            }
            SupervisionStrategy::RestForOne => {
                let (start, _, _) = self.launched.get(&id).ok_or(())?;
                let start = *start;

                self.restart(start..).await;
            }
        }

        Ok(())
    }
```
It will work, however, I'm not sure that will be correct because we're doing pop from the some place in the vector tracked by the supervisor and then pushing in the end. Is it ok for the `SupervisionStrategy::RestForOne` strategy?

**Comment by Relrin** [comment]
I have decided to change the signature of the restart/kill/stop method, so that it will use the `Range<T>` instead of the `RangeFrom<T>` type. After these changes, everything works as expected. :)

File: bastion/src/supervisor.rs [hunk 1]
```diff
@@ -1689,12 +1665,108 @@ impl Supervised {
...
+        self.max_restarts
+    }
+
+    /// Returns the current strategy for restarting failed actors.
+    pub fn strategy(&self) -> ActorRestartStrategy {
+        self.strategy.clone()
+    }
+
+    /// Sets the limit of attempts for restoring failed actors.
+    pub fn with_max_restarts(mut self, max_restarts: Option<usize>) -> Self {
```
Comment (vertexclique) [inline bastion/src/supervisor.rs]: I have surface API improvement:
* Let's make this isize
* isize = !0 means that unlimited restarts.
* isize > 0 means that use given restart count.
* isize = 0 means that don't restart.

What do you think @o0Ignition0o and @Relrin ?

File: bastion/examples/restart_strategy.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+use std::time::Duration;
+
+use bastion::prelude::*;
+
```
Comment (vertexclique) [inline bastion/examples/restart_strategy.rs]: We mostly write a prelude intro to our examples to explain where an example can be used for and what's the intent. I would like to have that. You might want to take a gaze to this one:
e.g. https://github.com/bastion-rs/bastion/blob/master/bastion/examples/parallel_computation.rs#L4

File: bastion/src/supervisor.rs [hunk 3]
```diff
@@ -1689,12 +1665,108 @@ impl Supervised {
...
+    /// # use std::time::Duration;
+    /// # use bastion::prelude::*;
+    /// #
+    /// # fn main() {
+    /// #     use std::time::Duration;
+    /// #     let actor_restart_strategy = ActorRestartStrategy::LinearBackOff {
+    /// #         timeout: Duration::from_secs(5)
+    /// #     };
+    /// #     let restart_strategy = RestartStrategy::default()
+    /// #        .with_actor_restart_strategy(actor_restart_strategy);
```
Comment (vertexclique) [inline bastion/src/supervisor.rs]: This just tests the initiation, that's nice but we need an actual test for this case in `tests/` I think:
https://github.com/bastion-rs/bastion/tree/master/bastion/tests

File: bastion/src/supervisor.rs (hunk 1 above)
Comment (Relrin) [reply bastion/src/supervisor.rs]: Perhaps it worth to reorganize it into enum (for better check in match expression), like this:
```
pub enum MaxRestartsPolicy {
    Unlimited,
    Limit(usize)
    DontRestart,
}
```
It looks better and easier to understand intentions of other devs from code.

Comment (o0Ignition0o) [reply bastion/src/supervisor.rs]: An enum sounds great
Maybe this naming would convey the meaning more accurately:
```rust
pub enum RestartPolicy {
    Always,
    Never,
    Tries(usize)
}

pub fn with_restarts(mut self, restarts: RestartPolicy) -> Self {}
```
What do you think ?

Comment (Relrin) [reply bastion/src/supervisor.rs]: Sounds good to me. I will refactor code with this enum

File: bastion/src/supervisor.rs (hunk 3 above)
Comment (Relrin) [reply bastion/src/supervisor.rs]: Done.

File: bastion/examples/restart_strategy.rs (hunk 2 above)
Comment (Relrin) [reply bastion/examples/restart_strategy.rs]: Done.

**Pull request merged**